### ISSUE TITLE
Allow applications to handle "not found" errors from `<resource>.get()`

### ISSUE THREAD

**Issue by clstokes** [pulumi/pulumi#3364]
### Problem Scenario
It can be common to need to query for an existing resource and if it doesn't exist then take some other action. Currently Pulumi's `<resource>.get()` methods halt Pulumi execution without throwing an error that the user/application can catch and react to.

### Suggestion
The results or error from `<resource>.get()` calls and data sources should be catchable so that users can take other actions as a result.

**Comment by clstokes** [comment]
Updated to include _"and data sources"_ in the suggestion.

**Comment by lukehoban** [comment]
> Updated to include "and data sources" in the suggestion.

For invokes ("data sources"), you can do this today:

```javascript
aws.getAmi({
    owners: ["153052954103"],
}, { async: true })
.then(
    res => console.log(`RESULT: ${res}`), 
    err => console.error(`ERROR: ${err}`)
);
```

That is, the error is bubbled up through the promise error handler (and as an exception in `async/await` code).

Another example:

```ts
async function getAmi(): Promise<string> {
    try {
        const amiDetails = await aws.getAmi({
            owners: ["153052954103"],
        }, { async: true });
        return amiDetails.id;
    } catch (err) {
        return "ami-014a2e30da708ee8b";
    }
}
```

This is *not* true for `.get()` calls though (which is what I will continue to track with this issue).

**Comment by lukehoban** [comment]
> Currently Pulumi's <resource>.get() methods halt Pulumi execution without throwing an error that the user/application can catch and react to.

This is true, and is deeply intertwined with the fact that these `.get()` calls (`ReadResource` in the engine and RPC) are serialized into the checkpoint file.   We treat them like all other CRUD operations on resources, and handle failures directly inside the engine, coordinating those failures with checkpointing of state.

It is likely that we do want to provide more user-program control over this, but it'll require a non-trivial update to make this happen.  Given that, we do not expect it to be part of 2.0, but will look at this as an opt-in thing post-2.0 and possibly a change of defaults for 3.0.

**Comment by VikramVasudevan** [comment]
+1 any updates on this one? Really need the get to be "catchable". The scenario I want to accomplish is as follows
- use .get() to lookup the resource. 
- if the resource does not exists, there is an exception thrown,  
- Catch the exception in the catch block and create a new resource there.

**Comment by 742617000027** [comment]
+1

Scenario: run a Lambda Function to create schema on an RDS Cluster only if the Cluster hasn't existed before.

**Comment by ValkyrieUK** [comment]
Expereincing the same issue. Would like to create an AWS Config recorder (limited to 1 per account per region) and want to check if one already exists before attempting to create one.

```ts
const recorder = aws.cfg.Recorder.get('recorder', 'default'); // Halts Pulumi if it does not exist and unable to catch the error.
```

**Comment by thomaszooman** [comment]
+1, please add error handling in `get(...)` or add `resource.exists(...)`

**Comment by napalm684** [comment]
Same situation with `aws.iot.ThingType.get()`. I need to be able to verify it exists before creating it.

**Comment by Matthew-2021** [comment]
Same issue here with getting custom domains registered to a CDN endpoint

**Comment by wrbbz** [comment]
Same issue here with an attempt to create k8s namespace.

Is there any progress on this one?

**Comment by Danielkiss9** [comment]
same issue while trying to check if an iam policy already exists.
any progress?

**Comment by t0yv0** [comment]
Found this documentation reference for `<resource>.get()`: https://www.pulumi.com/docs/intro/concepts/resources/get/

Would it be possible to add `<resource>.exists()` with a similar signature to `get()` with the following constraints:

- no exception if resource does not exists, return `false` (or `Promise<Boolean>`, or `Output<Boolean>`)

- perhaps avoid serializing `exists` calls into the checkpoint file and avoid the issues with that

- this is a non-breaking addition to the API 

The user code would then be able to express patterns like this:

```
if RDSCluster.exists(...) {
   cluster = RDSCluster.get(...)
} else {
   cluster = new RDSCluster(...)
   createSchema(cluster.getConn())
}
```

This might cause some confusion as to how the program behaves in `preview`, specifically:

- would the same code branch always be selected in `preview` and `up` against the same state?

- would `pulumi up` result in empty `pulumi preview`? this might no longer be true for programs that switch to a different branch after a `pulumi up` and might be a tad confusing; this seems inevitable though with any solution that allows differential behavior asked for in the ticket

**Comment by tejaswinibussety** [comment]
+1 any updates on this one?

**Comment by johan-vromo** [comment]
+1 any updates?

**Comment by starleaffff** [comment]
+1 for this.

**Comment by miabot-11** [comment]
+1

Any updates on this one? I am having a similar issue with `getPrivateZoneOutput()`

**Comment by t0yv0** [comment]
One more possible workaround worth mentioning for scenarios that involve managing a shared "one per account" resource such as AWS Config recorder is to use Stack References. The solution is to provision the Config Recorder in a dedicated Pulumi stack, and import it via Stack References into any stack that needs its.

**Comment by t0yv0** [comment]
Closing this off for now. Please let us know if you have a concrete use-case scenario we missed in the analysis and we need to reconsider.

**Comment by 1oglop1** [comment]
I have a use-case to use 2 different `get_resource` functions. I'd like to have `rds` database object, however RDS has 2 types `cluster` (aurora) and `instance` (classic rds) they both have almost the same methods and properties however AWS API differs with "Cluster" vs "DBinstance" so when querying for the database I use `rds.get_cluster` and it returned SDK error with the exception type `DBClusterNotFoundFault`, however some change the exception is different
```
Exception: invoke of aws:rds/getCluster:getCluster failed: invocation of aws:rds/getCluster:getCluster returned an error: invoking aws:rds/getCluster:getCluster: 1 error occurred:
        * reading RDS Cluster (myrds-db): couldn't find resource
```


before
```python
# db_params- is an object coming from a StackReference from another stack

try:
    existing_db = rds.get_cluster(
        cluster_identifier=db_params.apply(lambda x: x["db"]["identifier"]),
        opts=pulumi.InvokeOptions(provider=ec1_provider),
    )
except Exception as exc:
    if "DBClusterNotFoundFault" not in str(exc):
        raise exc
    existing_db = rds.get_instance(
        db_instance_identifier=db_params.apply(lambda x: x["db"]["identifier"]),
        opts=pulumi.InvokeOptions(provider=ec1_provider),
    )
```

after
```python
# db_params- is an object coming from a StackReference from another stack

try:
    existing_db = rds.get_cluster(
        cluster_identifier=db_params.apply(lambda x: x["db"]["identifier"]),
        opts=pulumi.InvokeOptions(provider=ec1_provider),
    )
except Exception as exc:
    if "couldn't find resource" not in str(exc):
        raise exc
    existing_db = rds.get_instance(
        db_instance_identifier=db_params.apply(lambda x: x["db"]["identifier"]),
        opts=pulumi.InvokeOptions(provider=ec1_provider),
    )
```

It would be great if I could the exception directly instead of `except Exception` it would be `except DBClusterNotFoundFault` or something like that

**Comment by Frassle** [comment]
Give the above & https://github.com/pulumi/pulumi/issues/11275 I think we should reopen this. We'll need to be careful to explain that this _can not_ be used for "try get or create" logic, but it sounds like it has other applications as well.

**Comment by ahilmathew** [comment]
Is there any way to lookup for the existence of a resource without the function actually throwing an error?
For context - I am trying to get `getUserAssignedIdentityOutput` but it errors out and the error seems to be not catchable. 
https://www.pulumi.com/registry/packages/azure-native/api-docs/managedidentity/getuserassignedidentity/

**Comment by dpunkturban** [comment]
+1 for that

**Comment by taxdorff-daimler** [comment]
Similar issue here with Confluent Users (Can't create an invite if they have an account, can't lookup a user if they don't) need try/catch to function:

`       try:
            inv = ccloud.Invitation(resource_name=name, email=config['email'], opts=ResourceOptions(retain_on_delete=True))
            id_val = inv.users.apply(lambda users: users[0].id)
        except:
            user = ccloud.get_user_output(email=config['email'])
            id_val = user.id.apply(lambda id: id)`

**Comment by rexebin** [comment]
A workaround would be to use Azure SDK to check if the resource exists before using the Get method.

**Comment by xfoxfu** [comment]
I have a similar use case which cannot be covered with "create if not exists" pattern.

```ts
```

**Comment by SnoozeFreddo** [comment]
+1

**Comment by alextricity25** [comment]
I have the same use case as @xfoxfu . My Pulumi program runs the `Release` resource to install a helm chart which creates a secret that I need to retrieve. I use the `Secret.get()` function, but now my preview keeps failing because that secret doesn't exist yet. I tried using `dependsOn: <myReleaseResource>` on that `get()` function, but no avail.

**Comment by mikocot** [comment]
> After an investigation in #10883 and discussing the implications with @Frassle we are leaning to close this issue and recommend other Pulumi mechanisms for the use cases here:
> 
> 1. A lot of the readers coming here are interested in handling the error from get to create a resource if it does not already exist.  [Support declaring that a resource should be created only if it does not already exist #3388](https://github.com/pulumi/pulumi/issues/3388) will provide a much better way to address this scenario.
> 2. There is a workaround available now before [Support declaring that a resource should be created only if it does not already exist #3388](https://github.com/pulumi/pulumi/issues/3388) ships. Providers expose data sources allowing the program to query for existence of resources in the cloud. For an example see [getSecurityGroups](https://www.pulumi.com/registry/packages/aws/api-docs/ec2/getsecuritygroups/) in AWS. The advantage of these compared to `Resource.get()` is that they do not interfere with Pulumi state.
> 3. For use cases that need to run some logic when a resource actually gets created, lifecycle hooks (specifically create hooks) as specified in [Support custom logic in resource lifecycle #1691](https://github.com/pulumi/pulumi/issues/1691) seem to be the preferable mechanism.

@t0yv0 @Frassle This is not a complete list unfortunately. For almost a year we're struggling with different instances of timing/consistency issues around cosmos db role assignments, where Pulumi successfully creates required identities but it's not yet propagated within Azure for CosmosDB to find it to performa assignment.

The workaround for that problem that we've been offered by Pulumi was exactly to use one of the above methods to wait until the resource is available and only then let Pulumi continue the deployment. Unfortunately even if we can complete the deployment in one go that way it is always considered a failed deployment as this methods reports errors for some iterations.

Hence, this issue not being solved makes the only workaround for that problem unviable. We were forced to go around pulumi in this case and use ArmClient directly, but one of the reasons we use Pulumi i exactly so we don't need to do that. It's also taken a lot of time to find this underlying issue, as it's simply not consistent nor expectable, and debugging options are limited. Therefore IMO you should rethink correcting it.

**Comment by marvkis** [comment]
I had a very similar 'user story':
* I triggered a Kubernetes operator via a CRD to do its job.
* It creates a whole a bunch of resources, including some secrets.
* In order for Pulumi to continue, I need the certificate of those secrets to properly configure a plugin. It needs the certificate to authenticate with the new instance created by the operator.

My current workaround - waiting several minutes before attempting to read the resource - is not very cool and often adds  unnecessary delay to the enrollment process.

**Comment by cooervo** [comment]
no fix for this after 5 years?

**Comment by joeduffy** [comment]
I never understood fully why this can't be fixed. I spent a few minutes just now, and though it's tricky, I do think there are a combination of changes we could make to address this:

1. Enable `ReadResource` to fail gracefully when a resource is missing. Right now, it just errors out in the step execution, which triggers a failure in the engine. This requires some thought as to the gRPC interface. The following is a clunky targeted change, but it works:

    ```diff
    diff --git a/pkg/resource/deploy/source_eval.go b/pkg/resource/deploy/source_eval.go
    index 9e890bdf7..e4025cdd8 100644
    --- a/pkg/resource/deploy/source_eval.go
    +++ b/pkg/resource/deploy/source_eval.go
    @@ -1381,8 +1381,14 @@ func (rm *resmon) ReadResource(ctx context.Context,
     		logging.V(5).Infof("ResourceMonitor.ReadResource operation canceled, name=%s", name)
     		return nil, rpcerror.New(codes.Unavailable, "resource monitor shut down while waiting on step's done channel")
 	    }
    -
     	contract.Assertf(result != nil, "ReadResource operation returned a nil result")
    +
    +	// If the state is not nil, which means it was found, marshal the properties and return them.
    +	if result.State.Outputs == nil {
    +		return nil, rpcerror.New(codes.InvalidArgument,
    +			fmt.Sprintf("resource %s of type %s with ID %s was not found", name, t, id))
    +	}
    +
     	marshaled, err := plugin.MarshalProperties(result.State.Outputs, plugin.MarshalOptions{
     		Label:            label,
     		KeepUnknowns:     true,
    @@ -1393,7 +1399,6 @@ func (rm *resmon) ReadResource(ctx context.Context,
     	if err != nil {
 	    	return nil, fmt.Errorf("failed to marshal %s return state: %w", result.State.URN, err)
     	}
    -
     	return &pulumirpc.ReadResourceResponse{
     		Urn:        string(result.State.URN),
     		Properties: marshaled,
    diff --git a/pkg/resource/deploy/step.go b/pkg/resource/deploy/step.go
    index a01afab1f..d1626dfef 100644
    --- a/pkg/resource/deploy/step.go
    +++ b/pkg/resource/deploy/step.go
    @@ -887,8 +887,10 @@ func (s *ReadStep) Apply() (resource.Status, StepCompleteFunc, error) {
     
     		// If there is no such resource, return an error indicating as such.
     		if result.Outputs == nil {
    -			return resource.StatusOK, nil, fmt.Errorf("resource '%s' does not exist", id)
    +			s.new.Outputs = nil
    +			return resourceStatus, func() { s.event.Done(&ReadResult{State: s.new}) }, nil
     		}
    +
     		s.new.Outputs = result.Outputs
     
     		if result.ID != "" {
    ```

    Most likely we'd want to think about an opt-in to this new failure mode and/or returning something in the response payload rather than abusing `codes.InvalidArgument` in this way. The engine also needs to "forget" a resource that triggers this behavior.

2. Adding a new variant of `get` methods -- call it `getSync` -- that returns a promise which breaks when the resource wasn't found. For example, we have `get(...): Bucket`, and this new method would be `getSync(...): Promise<Bucket>`.

3. That method internally just does what `get` does today, except that it internally `await`s the resource's `promise()`.

4. There were quite a few places in the language SDK where we assume that if a failure occurs in this manner, we want to terminate the program. That's probably an argument for a first class property in the `ReadResource` result, rather than abusing gRPC errors. We, of course, still want to fail _most of the time_, but not when `getSync` observed and communicated the error.

I was able to hack together a quick proof of concept, and indeed I was able to write this program:

```typescript
import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";
import * as awsx from "@pulumi/awsx";

// Create an AWS resource (S3 Bucket)
console.log("Before getSync");
try {
    const bucket = await aws.s3.BucketV2.getSync("my-bucket", "a-bucket-that-...-doesnt-exist");
    console.log("After getSync");
} catch (e) {
    console.log("Caught and ignoring: " + e);
}
```

And have it run w/out failure:

```
$ pulumi up
Previewing update (dev)

     Type                 Name               Plan       Info
 +   pulumi:pulumi:Stack  catchable-get-dev  create     2 messages

Diagnostics:
  pulumi:pulumi:Stack (catchable-get-dev):
    Before getSync
    Caught and ignoring: failed to read resource #a-bucket-that-...-doesnt-exist 'my-bucket' [aws:s3/bucketV2:BucketV2]: 3 INVALID_ARGUMENT: resource my-bucket of type aws:s3/bucketV2:BucketV2 with ID a-bucket-that-...-doesnt-exist was not found

Resources:
    + 1 to create

Do you want to perform this update?
> yes
  no
  details
```

The pattern of create-if-not-exist would presumably want to allocate a `"my-bucket"` in the catch, but that doesn't work without deeper changes alluded to above, such as forgetting about a `ReadResource` if it wasn't found during lookup. The code suggests this was the original intent, but it doesn't seem to do so.

**Comment by chrrlesWork** [comment]
> I have a similar use case which cannot be covered with "create if not exists" pattern.
> ...
> I have a database deployed with helm, and I have to read out the secrets created by the chart to get the connection password (which is then passed to other resources I wanted to create).

My use-case is similar.  I am using an AWS lambda to create certificates and insert them into the secrets manager.  One solution may be to pre-create placeholder secret entries, but that "doesn't feel right".  It would be ideal if there were a mechanism in Pulumi to mark resources which may not exist at time of preview.

Thanks Pulumi team!

**Comment by MeydanOzeri** [comment]
any update on this ? 

Very frustrating for use cases such as mine, i dynamically calculate storage needs for certain PVC's and if i see i need to expand storage i apply a patch on the PVC.

I only want to apply the patch if the size of the storage is different then what is in the cluster.
The issue is that i need to get the resource and inspect it's size, when i run this before the resource is created, Pulumi CLI fails and requires me to comment out the code, and after that to uncomment it and run again.

**Comment by chrrlesWork** [comment]
Bumping this issue to stay visible.  I am currently working around this issue by commenting out code for initial runs of IaC.

**Comment by naomatheus** [comment]
Seems like @joeduffy had a good suggestion in this comment:
 https://github.com/pulumi/pulumi/issues/3364#issuecomment-2496258371 

Can we get this implemented please? Currently dealing with a very simple use case, but cannot handle exceptions in a pythonic way. It seems like a quirk of Python, but with the popularity of pulumi's python interface this should be a priority, no?

**Comment by jdoupe** [comment]
Still a thing.  I'll reiterate the use case outlined in https://github.com/pulumi/pulumi-kubernetes/issues/1050 .   It seems like it'd be super common.  Monitoring resources such as `ServiceMonitor` and `VM*Scrape` cannot be created unless the underlying CRDs have been installed.  These are often installed via separate means.

**Comment by nb-ccd** [comment]
Another use case for this is wanting to manage database users with Pulumi in a program which can either:

- Create a database from a snapshot as part of a blue/green deployment (in which case, application users already exist within the database schema and I want to import them as Pulumi resources)
- Create an empty database for a fresh application stack

At the moment since there is no PostgreSQL role datasource or corresponding Pulumi `getRoles` method, so to make this work I think I'd need the input to the program to explicitly state whether users should be created or imported. This is inconvenient since I'd like the program to automatically create or import some users without explicitly configuration - eg for monitoring and for developer read access. It would be really nice to have the option to handle that conditional within the Pulumi program itself.

Sideway-related - the idea of "incremental import" of resources seems like it could provide an alternative path to handle some cases like this, but relies on providers exposing data sources: https://github.com/pulumi/pulumi/issues/17661
